components/net/http_loader.rs

```diff
@@ -1481,79 +1477,83 @@ async fn http_network_or_cache_fetch(
 
     // TODO(#33616) Step 8.22 If there’s a proxy-authentication entry, use it as appropriate.
 
-    block_for_cache_ready(
-        context,
-        http_request,
-        done_chan,
-        &mut revalidating_flag,
-        &mut response,
-    );
-
-    wait_for_cached_response(done_chan, &mut response).await;
+    {
+        // Section for blocking
+        let mut cache_guard = block_for_cache_ready(
+            context,
+            http_request,
+            done_chan,
+            &mut revalidating_flag,
+            &mut response,
+        )
+        .await;
 
-    // TODO(#33616): Step 9. If aborted, then return the appropriate network error for fetchParams.
+        wait_for_cached_response(done_chan, &mut response).await;
 
-    // Step 10. If response is null, then:
-    if response.is_none() {
-        // Step 10.1 If httpRequest’s cache mode is "only-if-cached", then return a network error.
-        if http_request.cache_mode == CacheMode::OnlyIfCached {
-            // The cache will not be updated,
-            // set its state to ready to construct.
-            update_http_cache_state(context, http_request);
-            return Response::network_error(NetworkError::Internal(
-                "Couldn't find response in cache".into(),
-            ));
-        }
+        let debug_request = http_request.clone();
+        // TODO(#33616): Step 9. If aborted, then return the appropriate network error for fetchParams.
 
-        // Step 10.2 Let forwardResponse be the result of running HTTP-network fetch given httpFetchParams,
-        // includeCredentials, and isNewConnectionFetch.
-        let forward_response =
-            http_network_fetch(http_fetch_params, include_credentials, done_chan, context).await;
-
-        let http_request = &mut http_fetch_params.request;
-        // Step 10.3 If httpRequest’s method is unsafe and forwardResponse’s status is in the range 200 to 399,
-        // inclusive, invalidate appropriate stored responses in httpCache, as per the
-        // "Invalidating Stored Responses" chapter of HTTP Caching, and set storedResponse to null.
-        if forward_response.status.in_range(200..=399) && !http_request.method.is_safe() {
-            let mut http_cache = context.state.http_cache.write();
-            http_cache.invalidate(http_request, &forward_response);
-        }
+        // Step 10. If response is null, then:
+        if response.is_none() {
+            // Step 10.1 If httpRequest’s cache mode is "only-if-cached", then return a network error.
+            if http_request.cache_mode == CacheMode::OnlyIfCached {
+                // The cache will not be updated,
+                // set its state to ready to construct.
```

Comment:
This comment is now out of date, it should just read: "Exit critical section on cache entry".